### PR TITLE
Align pos enum

### PR DESCRIPTION
**Pull request by BucketHatGuy**
hi guys

### CODE REVIEW

File: src/main/java/frc/robot/AlignPosition.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            AmpPos(),
+            SpeakerPos(),
+            LeftSourcePos(),
+            RightSourcePos(),
+            ClimbPos();
+
+            private static AlignPosition alignPosition;
+            public static Pose2d alignPose;
+
+            public static AlignPosition getPosition(){
```
Comment (notrandomness) [inline src/main/java/frc/robot/AlignPosition.java:18]: We also need to check if alignPosition is null, and then set it if it is.

File: src/main/java/frc/robot/AlignPosition.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                            alignPose = isRed? new Pose2d(-7, -3.3, Rotation2d.fromDegrees(-145)): new Pose2d(7.8, -3.8, Rotation2d.fromDegrees(-55));
+                            break;
+                        case RightSourcePos:
+                            alignPose = isRed? new Pose2d(-7.9, -3.8, Rotation2d.fromDegrees(-145)): new Pose2d(6.95, -3.3, Rotation2d.fromDegrees(-55));
+                            break;
+                        case ClimbPos:
+                            alignPose = isRed? new Pose2d(3.4, 0, null): new Pose2d(-3.4, 0, null);
+                            break;
+                    }
+                    return AlignPosition.SpeakerPos;
```
Comment (notrandomness) [inline src/main/java/frc/robot/AlignPosition.java]: This return line should be removed.

File: src/main/java/frc/robot/AlignPosition.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+public enum AlignPosition {
+            AmpPos(),
+            SpeakerPos(),
+            LeftSourcePos(),
+            RightSourcePos(),
+            ClimbPos();
+
+            private static AlignPosition alignPosition;
+            public static Pose2d alignPose;
```
Comment (notrandomness) [inline src/main/java/frc/robot/AlignPosition.java]: I think we should make this private and add a getter for it.

File: src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -149,6 +154,34 @@ public static final class AutoConstants { //TODO: The below constants are used i
         public static final TrapezoidProfile.Constraints kThetaControllerConstraints =
             new TrapezoidProfile.Constraints(
                 kMaxAngularSpeedRadiansPerSecond, kMaxAngularSpeedRadiansPerSecondSquared);
+
+        public static Pair<Double, Double> alignCoordinates;
+
+        public enum AlignPosition {
```
Comment (notrandomness) [inline src/main/java/frc/robot/Constants.java]: I don't think we need this, you might just revert the changes in this file.

File: src/main/java/frc/robot/commands/DefaultTeleop.java [hunk 5]
```diff
@@ -26,41 +21,43 @@ public class DefaultTeleop extends Command{
...
+        this.driver = driver;
+        this.operator = operator;
+        translationSup = XboxController.Axis.kLeftY.value;
+        strafeSup = XboxController.Axis.kLeftX.value;
+        rotationSup = XboxController.Axis.kRightX.value;
+        robotCentricSup = true;
+        bButtonPressed = buttonDebouncer.calculate(driver.getRawButtonReleased(XboxController.Button.kB.value));
+        aButtonPressed = buttonDebouncer.calculate(driver.getRawButtonReleased(XboxController.Button.kA.value));
+        xButtonPressed = buttonDebouncer.calculate(driver.getRawButtonReleased(XboxController.Button.kX.value));
+        yButtonPressed = buttonDebouncer.calculate(driver.getRawButtonReleased(XboxController.Button.kY.value));
```
Comment (notrandomness) [inline src/main/java/frc/robot/commands/DefaultTeleop.java]: buttonDebouncer.calculate(driver.getRawButtonReleased(...) is just returning false, right? This is in the constructor, so it's only called once when the robot starts. We should be initializing these button variables to false here.

File: src/main/java/frc/robot/commands/DefaultTeleop.java [hunk 6]
```diff
@@ -88,19 +85,21 @@ public void execute() {
...
 
-        if ((MathUtil.applyDeadband(rotationAxis, Constants.stickDeadband) > 0 && isAligning) || (bButtonPressed && isAligning)) {
+        if ((isAligning && MathUtil.applyDeadband(rotationAxis, Constants.stickDeadband) != 0) || isAligning && ((bButtonPressed || xButtonPressed || yButtonPressed || aButtonPressed))) {
             isAligning = false;
-        } else if (bButtonPressed && isAligning == false) {
+        } else if (!isAligning && (bButtonPressed || xButtonPressed || yButtonPressed || aButtonPressed)) {
             isAligning = true;
         }
 
+        if (bButtonPressed) {
```
Comment (notrandomness) [inline src/main/java/frc/robot/commands/DefaultTeleop.java]: We're never changing the value of bButtonPressed or any of the other button variables. We need to be checking if these buttons were pressed and updating these values in execute. They will also need separate debouncers, cause calling the same debouncer multiple times per iteration will invalidate its internal values.

File: src/main/java/frc/robot/AlignPosition.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public enum AlignPosition {
+            AmpPos(),
+            SpeakerPos(),
+            LeftSourcePos(),
+            RightSourcePos(),
+            ClimbPos();
```
Comment (notrandomness) [inline src/main/java/frc/robot/AlignPosition.java]: We should add an option for manual control (or no alignment). Can probably set the alignPose to NULL in that case.

File: src/main/java/frc/robot/commands/DefaultTeleop.java [hunk 8]
```diff
@@ -88,19 +85,21 @@ public void execute() {
...
-        } else if (MathUtil.applyDeadband(rotationAxis, Constants.stickDeadband) > 0) {
-            isAligning = false;
-        }
 
-        if ((MathUtil.applyDeadband(rotationAxis, Constants.stickDeadband) > 0 && isAligning) || (bButtonPressed && isAligning)) {
+        if ((isAligning && MathUtil.applyDeadband(rotationAxis, Constants.stickDeadband) != 0) || isAligning && ((bButtonPressed || xButtonPressed || yButtonPressed || aButtonPressed))) {
             isAligning = false;
-        } else if (bButtonPressed && isAligning == false) {
+        } else if (!isAligning && (bButtonPressed || xButtonPressed || yButtonPressed || aButtonPressed)) {
             isAligning = true;
```
Comment (notrandomness) [inline src/main/java/frc/robot/commands/DefaultTeleop.java]: All of this isAligning and buttonPressed logic can be simplified by using the AlignPosition enum. In fact, I don't think we need the isAligning variable at all, we can probably just remove this if/else if block.

File: src/main/java/frc/robot/commands/DefaultTeleop.java [hunk 9]
```diff
@@ -88,19 +85,21 @@ public void execute() {
...
 
+        if (bButtonPressed) {
+            AlignPosition.setPosition(AlignPosition.AmpPos);
+        } else if (aButtonPressed) {
+            AlignPosition.setPosition(AlignPosition.ClimbPos);
+        } else if (xButtonPressed) {
+            AlignPosition.setPosition(AlignPosition.SpeakerPos);
+        } // find a way to bind each source side to a button scheme that works
+
         rotationVal = (isAligning? s_AutoRotateUtil.calculateRotationSpeed():MathUtil.applyDeadband(rotationAxis, Constants.stickDeadband));
```
Comment (notrandomness) [inline src/main/java/frc/robot/commands/DefaultTeleop.java]: Instead of checking isAligning, we can just check if AlignPosition.equals(AlignPosition.Manual) or something to that effect.

File: src/main/java/frc/robot/commands/DefaultTeleop.java [hunk 10]
```diff
@@ -88,19 +85,21 @@ public void execute() {
...
-        } else if (bButtonPressed && isAligning == false) {
+        } else if (!isAligning && (bButtonPressed || xButtonPressed || yButtonPressed || aButtonPressed)) {
             isAligning = true;
         }
 
+        if (bButtonPressed) {
+            AlignPosition.setPosition(AlignPosition.AmpPos);
+        } else if (aButtonPressed) {
+            AlignPosition.setPosition(AlignPosition.ClimbPos);
+        } else if (xButtonPressed) {
```
Comment (notrandomness) [inline src/main/java/frc/robot/commands/DefaultTeleop.java]: We really only need to call setPosition once per button press. After that, the button variable can be set back to false. I recommend moving this logic into a separate method (along with the button updating logic) to make this cleaner.

**Review by notrandomness — CHANGES_REQUESTED** [review]

**Comment by notrandomness** [comment]
I would also recommend adding these changes from the DefaultTeleop branch while you're at it: https://github.com/CSE-FortOsageCTC/5801_Inspire_2024/commit/b101599140604698366d435ad4baff67506354d9

And this from the DriveUpdates branch: https://github.com/CSE-FortOsageCTC/5801_Inspire_2024/commit/d17ac524c6355c9eb278ae51f1bbad036d465d51

File: src/main/java/frc/robot/AlignPosition.java (hunk 2 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/AlignPosition.java]: Done

File: src/main/java/frc/robot/Constants.java (hunk 4 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/Constants.java]: goodbye alignposition enum 👋

File: src/main/java/frc/robot/commands/DefaultTeleop.java (hunk 5 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/commands/DefaultTeleop.java]: added buttonDebouncer.calculate(.... in execute before they are used and replaced the ones in the constructor by setting them to false

File: src/main/java/frc/robot/commands/DefaultTeleop.java (hunk 6 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/commands/DefaultTeleop.java]: Values should now update in execute and made 3 more debouncers

File: src/main/java/frc/robot/AlignPosition.java (hunk 7 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/AlignPosition.java]: added extra options, "Manual()", and "AutoPickup()" and Manual case and AutoPickup case

File: src/main/java/frc/robot/commands/DefaultTeleop.java (hunk 8 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/commands/DefaultTeleop.java]: removed ALL of the previous "isButtonPressed" items due to realizing we can just call the button switches in RobotContainer now. removed button debouncers, removed isAligning boolean. Added if/else block setting the rotation value to the joystick if the enum is Manual and uses the pid loop if not. Added logic in RobotContainer to switch Robot Align State

File: src/main/java/frc/robot/AlignPosition.java (hunk 3 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/AlignPosition.java]: boom done

File: src/main/java/frc/robot/AlignPosition.java (hunk 1 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/AlignPosition.java:18]: now returns SpeakerPos as default if null

File: src/main/java/frc/robot/commands/DefaultTeleop.java (hunk 9 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/commands/DefaultTeleop.java]: removed buttonPressed objects.... hehe

File: src/main/java/frc/robot/commands/DefaultTeleop.java (hunk 10 above)
Comment (T-Wezzy) [reply src/main/java/frc/robot/commands/DefaultTeleop.java]: removed already in previous comment... hehe

**Review by notrandomness — APPROVED** [review]
